### PR TITLE
python version conflict

### LINKED ISSUE #5: Exercise: incompatible python code
- Two containers, one in python2, one in python 3
Story:
- You need code written by another phd student who isn't available anymore. It's written in python2. Your code is in python 3

For python2: `print hi guys`
For python3: `print(f'hi {name}')`

Input and output should be shared with docker volumes

---

Thanks for contributing! If this contribution is for instructor training, please send an email to checkout@carpentries.org with a link to this contribution so we can record your progress. You’ve completed your contribution step for instructor checkout just by submitting this contribution.  

If this issue is about a specific episode within a lesson, please provide its link or filename.

Please keep in mind that lesson maintainers are volunteers and it may be some time before they can respond to your contribution. Although not all contributions can be incorporated into the lesson materials, we appreciate your time and effort to improve the curriculum.  If you have any questions about the lesson maintenance process or would like to volunteer your time as a contribution reviewer, please contact The Carpentries Team at team@carpentries.org.  

---

### PR DESCRIPTION
**Pull request by JensWehner**
closes #10

### CODE REVIEW

File: code/convertbirthdaytoage.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+
+import csv
+import json
+import datetime
+import os
```
Comment (JensWehner) [inline code/convertbirthdaytoage.py]:
```suggestion
```

**Review by JensWehner — COMMENTED** [review]
@rogerkuou That is fantastic thanks for the work. :)

File: _episodes/09-practical-examples.md [hunk 2]
```diff
@@ -7,5 +7,86 @@ questions:
...
+  - Docker can be a solution to include the legacy script into your workflow with minimum effort.
+  - You can use intermediate files to let different Docker containers talk to each other.
 ---
+
+You are a social science researcher and are trying to establish how many people in a list are below the allowed drinking age.
+You are provided with a list of names and the corresponding birthdates.
+
+A former PhD student already did the first part and wrote a program to convert birthdates into the current age of the people. Unfortunately it was written in Python2.7 and there is no way to upgrade it to Python3 (technically there is, but psssst).
+
+As you are on of the best and brightest you have written a code to check the current age of people against the legal drinking age and output all people who are below it. You wrote it in Python 3.6. Your professor/boss/overlord and the rest of the group use python 3.3 in the work and cannot use your critical piece of software.
```
Comment (dsmits) [inline _episodes/09-practical-examples.md]: ``
```suggestion
As you are on of the best and brightest you have written a script to check the current age of people against the legal drinking age and output all people who are below it. You wrote it in Python 3.6. Your professor/boss/overlord and the rest of the group use python 3.3 in the work and cannot use your critical piece of software.
```

File: _episodes/09-practical-examples.md [hunk 3]
```diff
@@ -7,5 +7,86 @@ questions:
...
+
+Create two docker containers and provide your colleagues with a working docker workflow before they get angry!
+
+## Preparation work
+
+Before you start with a solution, you need to download the files you need:
+
+- the old script [convertbirthdaytoage.py](/code/convertbirthdaytoage.py) written by the former PhD in Python 2.7;
+- the new script [heckdrinkinglimit.py](/code/checkdrinkinglimit.py) you wrote in Python 3.6;
```
Comment (dsmits) [inline _episodes/09-practical-examples.md]:
```suggestion
- the new script [checkdrinkinglimit.py](/code/checkdrinkinglimit.py) you wrote in Python 3.6;
```

File: _episodes/09-practical-examples.md [hunk 4]
```diff
@@ -7,5 +7,86 @@ questions:
 objectives:
   - Use docker in combination with real research software
 keypoints:
--  
+  - Docker can be a solution to include the legacy script into your workflow with minimum effort.
+  - You can use intermediate files to let different Docker containers talk to each other.
 ---
+
+You are a social science researcher and are trying to establish how many people in a list are below the allowed drinking age.
```
Comment (dsmits) [inline _episodes/09-practical-examples.md]: Could you add `>` and `> >` before the lines of the exercise so it is formatted properly?
Also, it needs to end with `{: .challenge}

Comment (dsmits) [reply _episodes/09-practical-examples.md]: Could you add a title to the exercise?

File: _episodes/09-practical-examples.md [hunk 5]
```diff
@@ -7,5 +7,86 @@ questions:
...
+{: .language-bash}
+
+~~~
+checkdrinkinglimit.py  convertbirthdaytoage.py  data_birthday.csv
+~~~
+{: .output}
+
+
+> ## Solution
+> Now let's think of our problem: we need to use the old script `convertbirthdaytoage.py` to get the age of people, and use the new script `heckdrinkinglimit.py` to find out if anyone is under the drinking age. Neither of the scripts is 100% compatible with your collegues' Python 3.3 environment. To make sure they have a smooth workflow, we need to run the old script under Python2.7 environment, and the new script under Python 3.6.
```
Comment (dsmits) [inline _episodes/09-practical-examples.md]:
```suggestion
> Now let's think of our problem: we need to use the old script `convertbirthdaytoage.py` to get the age of people, and use the new script `checkdrinkinglimit.py` to find out if anyone is under the drinking age. Neither of the scripts is 100% compatible with your colleagues' Python 3.3 environment. To make sure they have a smooth workflow, we need to run the old script under Python2.7 environment, and the new script under Python 3.6. 
```

File: _episodes/09-practical-examples.md [hunk 6]
```diff
@@ -7,5 +7,86 @@ questions:
...
+
+## Preparation work
+
+Before you start with a solution, you need to download the files you need:
+
+- the old script [convertbirthdaytoage.py](/code/convertbirthdaytoage.py) written by the former PhD in Python 2.7;
+- the new script [heckdrinkinglimit.py](/code/checkdrinkinglimit.py) you wrote in Python 3.6;
+- the [birthday data](/data/data_birthday.csv) required for this study.
+
+To simplify our workflow, let's put them all in the same working directory:
```
Comment (dsmits) [inline _episodes/09-practical-examples.md]: Is this something we need to clarify a bit more? Like, create a new folder and copy the files to it?

File: _episodes/09-practical-examples.md [hunk 7]
```diff
@@ -7,5 +7,86 @@ questions:
...
+> ~~~
+> {: .language-bash}
+>
+> After we have the images ready, we can run  `convertbirthdaytoage.py` to get the age of people:
+> ~~~
+> $ docker run -v $(pwd):/mnt -w /mnt python:2.7 python /mnt/convertbirthdaytoage.py
+> ~~~
+> {: .language-bash}
+>
+> Here we used the `-v $(pwd):/mnt` to mount our working directory to the Docker conatiner. `-w /mnt` is used to specify the working directory within the Docker container. Now if we check our current working directory:
```
Comment (dsmits) [inline _episodes/09-practical-examples.md]:
```suggestion
> Here we used the `-v $(pwd):/mnt` to mount our working directory to the Docker container. `-w /mnt` is used to specify the working directory within the Docker container. Now if we check our current working directory:
```

File: _episodes/09-practical-examples.md [hunk 8]
```diff
@@ -7,5 +7,86 @@ questions:
...
+> ~~~
+> $ ls .
+> ~~~
+> {: .language-bash}
+>
+> ~~~
+> ages.json  checkdrinkinglimit.py  convertbirthdaytoage.py  data_birthday.csv
+> ~~~
+> {: .output}
+> A new file `ages.json` has been created, which contains the ages of all people. Now we can execute `heckdrinkinglimit.py` using another Docker image:
```
Comment (dsmits) [inline _episodes/09-practical-examples.md]:
```suggestion
> A new file `ages.json` has been created, which contains the ages of all people. Now we can execute `checkdrinkinglimit.py` using another Docker image:
```

File: _episodes/09-practical-examples.md [hunk 9]
```diff
@@ -7,5 +7,86 @@ questions:
...
+> ~~~
+> $ docker run -v $(pwd):/mnt -w /mnt python:3.6 python /mnt/checkdrinkinglimit.py
+> ~~~
+> {: .language-bash}
+> ~~~
+> Daria is not allowed to drink because they are only 6 years old!
+> Sherilyn is not allowed to drink because they are only 8 years old!
+> Clint is not allowed to drink because they are only 14 years old!
+> Val is not allowed to drink because they are only 13 years old!
```
Comment (dsmits) [inline _episodes/09-practical-examples.md]: Some groundbreaking sociological results ;)

**Review by dsmits — CHANGES_REQUESTED** [review]
Some small remarks, if you've addressed them you can merge it.
Great work, I like the example that you made up.

File: _episodes/09-practical-examples.md (hunk 6 above)
Comment (JensWehner) [reply _episodes/09-practical-examples.md]: Actually that is a good question. I would actually prefer to add all the data to the zip file, but that makes version controlling it so much harder.

**Review by mkuzak — COMMENTED** [review]
I really like this example. I have some comments about the style. I will leave those inline.

File: _episodes/09-practical-examples.md [hunk 10]
```diff
@@ -7,5 +7,89 @@ questions:
...
 ---
+
+> ## Python Workflow
+>
+> You are a social science researcher and are trying to establish how many people in a list are below the allowed drinking age.
+> You are provided with a list of names and the corresponding birthdates.
+>
+> A former PhD student already did the first part and wrote a program to convert birthdates into the current age of the people. Unfortunately it was written in Python2.7 and there is no way to upgrade  it to Python3 (technically there is, but psssst).
+>
+> As you are on of the best and brightest you have written a script to check the current age of people against the legal drinking age and output all people who are below it. You wrote it in Python 3.6. Your professor/boss/overlord and the rest of the group use python 3.3 in the work and cannot use your critical piece of software.
```
Comment (mkuzak) [inline _episodes/09-practical-examples.md]:
```suggestion
> As you are one of the best and brightest you have written a script to check the current age of people against the legal drinking age and output all people who are below it. You wrote it in Python 3.6. Your professor or boss and the rest of the group use python 3.3 in the work and cannot use your useful piece of software.
```

File: _episodes/09-practical-examples.md [hunk 11]
```diff
@@ -7,5 +7,89 @@ questions:
...
+> 
+> Create two docker containers and provide your colleagues with a working docker workflow before they get angry!
```
Comment (mkuzak) [inline _episodes/09-practical-examples.md]:
```suggestion
> Create two docker containers and provide your colleagues with a working docker workflow.
```

File: _episodes/09-practical-examples.md [hunk 12]
```diff
@@ -7,5 +7,89 @@ questions:
...
+> ~~~
+> checkdrinkinglimit.py  convertbirthdaytoage.py  data_birthday.csv
+> ~~~
+> {: .output}
+>
+>
+> > ## Solution
+> > Now let's think of our problem: we need to use the old script `convertbirthdaytoage.py` to get the age of people, and use the new script `checkdrinkinglimit.py` to find out if anyone is under the drinking age. Neither of the scripts is 100% compatible with your colleagues' Python 3.3 environment. To make sure they have a smooth workflow, we need to run the old script under Python2.7 environment, and the new script under Python 3.6. 
+> >
+> > From what we have learnt, we know `DockerHub` provides images for both environments we need. What we need to do is simply pulling them down:
```
Comment (mkuzak) [inline _episodes/09-practical-examples.md]:
```suggestion
> > From what we have learnt, we know `DockerHub` provides images for both environments we need. What we need to do is to pull them down:
```

File: _episodes/09-practical-examples.md [hunk 13]
```diff
@@ -7,5 +7,89 @@ questions:
 objectives:
   - Use docker in combination with real research software
 keypoints:
--  
+  - Docker can be a solution to include the legacy script into your workflow with minimum effort.
+  - You can use intermediate files to let different Docker containers talk to each other.
 ---
+
+> ## Python Workflow
```
Comment (dsmits) [inline _episodes/09-practical-examples.md]:
```suggestion
> ## Python Version Conflict
```

**Review by dsmits — CHANGES_REQUESTED** [review]
Sorry to drag this PR on, just one suggestion for the title.

**Review by dsmits — APPROVED** [review]

**Pull request merged**